No var keyword rule

DickvdBrink opened the pull request:
Added no-var-keyword rule.

I couldn't think of a better name for this rule so if you have a better name I will change it :)
fix #378
Feel free to close this PR if this isn't the way you want it :)

Btw: I'm currently on vacation so feedback might take till around Sunday :)

ash211 commented on README.md:
```diff
@@ -172,6 +172,7 @@ A sample configuration file with all options is available [here](https://github.
     * `"check-parameters"` disallows unused function and constructor parameters.
         * NOTE: this option is experimental and does not work with classes that use abstract method declarations, among other things. Use at your own risk.
 * `no-use-before-declare` disallows usage of variables before their declaration.
+* `no-var-keyword` disallows ussage of the `var` keyword, use `let` or `const` instead.
```
Typo: ussage -> usage

adidahiya commented on src/rules/noVarKeywordRule.ts:
```diff
@@ -0,0 +1,71 @@
+
+        if (!this.skipVariableDeclaration) {
+            var flags = node.parent.flags;
+            var name = (<ts.Identifier> node.name);
```
`node.name` might be a `BindingPattern` (a.k.a. destructuring), so you'll have to handle both cases.

adidahiya commented on src/rules/noVarKeywordRule.ts:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public visitVariableDeclaration(node: ts.VariableDeclaration) {
+
```
remove this newline

adidahiya commented on src/rules/noVarKeywordRule.ts:
```diff
@@ -0,0 +1,71 @@
+                return true;
+            }
+        }
+        return false;
```
L62–L68 can be similified:

```ts
return modifiers.some((modifier) => modifier.kind === modifierKind);
```

adidahiya replied:
After you rebase on the latest `master`, you can use the new util function `Lint.hasModifier` instead.

adidahiya commented on src/rules/noVarKeywordRule.ts:
```diff
@@ -0,0 +1,71 @@
+
+    public visitVariableStatement(node: ts.VariableStatement): void {
+        if (this.hasModifier(node.modifiers, ts.SyntaxKind.ExportKeyword) ||
+            this.hasModifier(node.modifiers, ts.SyntaxKind.DeclareKeyword)) {
```
please move the `||` operator to the next line

adidahiya commented on test/files/rules/novarkeyword.test.ts:
```diff
@@ -0,0 +1,25 @@
+
+module quz {
+    export var i;
+}
```
let's add some more tests for 1.5 syntax, like `export default`, `import`, destructuring (`var [a, b] = [1, 2]`)...

adidahiya commented:
Any estimate on when you'll be able to make these fixes? Should I take up this PR?

DickvdBrink commented:
Oops sorry, will fix some stuff tonight. Actually forgot about this a bit...

I do have a small question though!
The location where this error is thrown might be a bit odd, it is now thrown twice on the code below and on the location of the name but there is only one `var` keyword so that might be a bit weird. Do you agree?

    var i,
        j;

adidahiya commented:
I agree, it's excessive to report 2 failures in the case of a variable declaration list of length 2 like that. You could try making the walker visit `VariableDeclarationList` nodes to get the desired behavior.

DickvdBrink commented:
Fixed most of the feedback stuff (hopefully in the right way) and rebased on master.

What I didn't fix (yet) is the `export default` stuff because I don't know how to make it usable in a single TypeScript file..
